Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean

```diff
@@ -223,3 +244,105 @@ lemma IsPrimitiveRoot.autToPow_eq_ModularCyclotomicCharacter (n : ℕ) [NeZero n
     Function.comp_apply, Units.coe_mapEquiv, MonoidHom.coe_toHomUnits, MonoidHom.coe_mk,
     OneHom.coe_mk, RingEquiv.coe_toMulEquiv, ZMod.ringEquivCongr_val, AlgEquiv.coe_ringEquiv]
     using ModularCyclotomicCharacter.toFun_spec'' g hμ
+
+/-
+
+## The p-adic theory
+
+-/
+
+open ModularCyclotomicCharacter in
+/-- The underlying function of the cyclotomic character. See `CyclotomicCharacter`. -/
+noncomputable def CyclotomicCharacter.toFun (p : ℕ) [Fact p.Prime] (g : L ≃+* L) : ℤ_[p] :=
+  letI := Classical.propDecidable
+  if H : ∀ (i : ℕ), ∃ ζ : L, IsPrimitiveRoot ζ (p ^ i) then
+    PadicInt.ofIntSeq _ (PadicInt.isCauSeq_padicNorm_of_pow_dvd_sub
+      (aux g <| p ^ ·) _ fun i ↦ pow_dvd_aux_pow_sub_aux_pow g p H _ _ i.le_succ)
+  else 1
+
+namespace CyclotomicCharacter
+
+local notation "χ" => CyclotomicCharacter.toFun
+
+variable (p : ℕ) [Fact p.Prime] (g : L ≃+* L) (H : ∀ i, ∃ ζ : L, IsPrimitiveRoot ζ (p ^ i))
```

Comment:
Does it make sense to introduce a class for `H` and ignore the cases when it doesn't hold?

Reply:
There seems to be a `HasEnoughRootsOfUnity` class. Not sure if I should but I've kept the if-then-else thing for now.